File: taurus-protect-sdk-typescript/src/mappers/exchange.ts
/**
 * Exchange mapper functions for converting OpenAPI DTOs to domain models.
 */

import type { Exchange, ExchangeCounterparty, ExchangeWithdrawalFee } from '../models/exchange';
import { currencyFromDto } from './currency';
import { safeBool, safeDate, safeMap, safeString } from './base';

/**
 * Maps an exchange DTO to an Exchange domain model.
 */
export function exchangeFromDto(dto: unknown): Exchange | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    id: safeString(d.id),
    exchange: safeString(d.exchange),
    account: safeString(d.account),
    currency: safeString(d.currency),
    type: safeString(d.type),
    totalBalance: safeString(d.totalBalance ?? d.total_balance),
    status: safeString(d.status),
    container: safeString(d.container),
    label: safeString(d.label),
    displayLabel: safeString(d.displayLabel ?? d.display_label),
    baseCurrencyValuation: safeString(d.baseCurrencyValuation ?? d.base_currency_valuation),
    hasWLA: safeBool(d.hasWLA ?? d.has_wla),
    currencyInfo: currencyFromDto(d.currencyInfo ?? d.currency_info),
    creationDate: safeDate(d.creationDate ?? d.creation_date),
    updateDate: safeDate(d.updateDate ?? d.update_date),
  };
}

/**
 * Maps an array of exchange DTOs to Exchange domain models.
 */
export function exchangesFromDto(dtos: unknown[] | null | undefined): Exchange[] {
  return safeMap(dtos, exchangeFromDto);
}

/**
 * Maps an exchange counterparty DTO to an ExchangeCounterparty domain model.
 */
export function exchangeCounterpartyFromDto(dto: unknown): ExchangeCounterparty | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  return {
    name: safeString(d.name),
    baseCurrencyValuation: safeString(d.baseCurrencyValuation ?? d.base_currency_valuation),
  };
}

/**
 * Maps an array of exchange counterparty DTOs to ExchangeCounterparty domain models.
 */
export function exchangeCounterpartiesFromDto(
  dtos: unknown[] | null | undefined
): ExchangeCounterparty[] {
  return safeMap(dtos, exchangeCounterpartyFromDto);
}

/**
 * Maps a withdrawal fee response to an ExchangeWithdrawalFee domain model.
 */
export function exchangeWithdrawalFeeFromDto(dto: unknown): ExchangeWithdrawalFee | undefined {
  if (!dto || typeof dto !== 'object') {
    return undefined;
  }

  const d = dto as Record<string, unknown>;
  // The API returns the fee as 'result' in the reply
  const fee = safeString(d.result ?? d.fee);

  if (fee === undefined) {
    return undefined;
  }

  return {
    fee,
  };
}

File: taurus-protect-sdk-go/pkg/protect/mapper/exchange.go
package mapper

import (
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/internal/openapi"
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/model"
)

// ExchangeFromDTO converts an OpenAPI TgvalidatordExchange to a domain Exchange.
func ExchangeFromDTO(dto *openapi.TgvalidatordExchange) *model.Exchange {
	if dto == nil {
		return nil
	}

	exchange := &model.Exchange{
		ID:                    safeString(dto.Id),
		Exchange:              safeString(dto.Exchange),
		Account:               safeString(dto.Account),
		Currency:              safeString(dto.Currency),
		Type:                  safeString(dto.Type),
		TotalBalance:          safeString(dto.TotalBalance),
		Status:                safeString(dto.Status),
		Container:             safeString(dto.Container),
		Label:                 safeString(dto.Label),
		DisplayLabel:          safeString(dto.DisplayLabel),
		HasWLA:                safeBool(dto.HasWLA),
		BaseCurrencyValuation: safeString(dto.BaseCurrencyValuation),
	}

	// Convert timestamps
	if dto.CreationDate != nil {
		exchange.CreationDate = *dto.CreationDate
	}
	if dto.UpdateDate != nil {
		exchange.UpdateDate = *dto.UpdateDate
	}

	// Convert currency info
	if dto.CurrencyInfo != nil {
		exchange.CurrencyInfo = CurrencyFromDTO(dto.CurrencyInfo)
	}

	return exchange
}

// ExchangesFromDTO converts a slice of OpenAPI TgvalidatordExchange to domain Exchanges.
func ExchangesFromDTO(dtos []openapi.TgvalidatordExchange) []*model.Exchange {
	if dtos == nil {
		return nil
	}
	exchanges := make([]*model.Exchange, len(dtos))
	for i := range dtos {
		exchanges[i] = ExchangeFromDTO(&dtos[i])
	}
	return exchanges
}

// ExchangeCounterpartyFromDTO converts an OpenAPI TgvalidatordExchangeCounterparty to a domain ExchangeCounterparty.
func ExchangeCounterpartyFromDTO(dto *openapi.TgvalidatordExchangeCounterparty) *model.ExchangeCounterparty {
	if dto == nil {
		return nil
	}

	return &model.ExchangeCounterparty{
		Name:                  safeString(dto.Name),
		BaseCurrencyValuation: safeString(dto.BaseCurrencyValuation),
	}
}

// ExchangeCounterpartiesFromDTO converts a slice of OpenAPI TgvalidatordExchangeCounterparty to domain ExchangeCounterparties.
func ExchangeCounterpartiesFromDTO(dtos []openapi.TgvalidatordExchangeCounterparty) []*model.ExchangeCounterparty {
	if dtos == nil {
		return nil
	}
	counterparties := make([]*model.ExchangeCounterparty, len(dtos))
	for i := range dtos {
		counterparties[i] = ExchangeCounterpartyFromDTO(&dtos[i])
	}
	return counterparties
}
